src/envs/tools.py

```diff
@@ -0,0 +1,643 @@
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._Place = ParameterizedOption(
+            # variables: [robot]
+            # params: [absolute x, absolute y]
+            "Place",
+            types=[self._robot_type],
+            params_space=Box(
+                np.array([self.table_lx, self.table_ly], dtype=np.float32),
+                np.array([self.table_ux, self.table_uy], dtype=np.float32)),
+            _policy=self._Place_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenScrewWithScrewdriver = ParameterizedOption(
+            # variables: [robot, screw, screwdriver]
+            # params: []
+            "FastenScrewWithScrewdriver",
+            types=[
+                self._robot_type, self._screw_type, self._screwdriver_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenScrewByHand = ParameterizedOption(
+            # variables: [robot, screw]
+            # params: []
+            "FastenScrewByHand",
+            types=[self._robot_type, self._screw_type, self._contraption_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenNailWithHammer = ParameterizedOption(
+            # variables: [robot, nail, hammer]
+            # params: []
+            "FastenNailWithHammer",
+            types=[
+                self._robot_type, self._nail_type, self._hammer_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._FastenBoltWithWrench = ParameterizedOption(
+            # variables: [robot, bolt, wrench]
+            # params: []
+            "FastenBoltWithWrench",
+            types=[
+                self._robot_type, self._bolt_type, self._wrench_type,
+                self._contraption_type
+            ],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Fasten_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        # Objects
+        self._robot = Object("robby", self._robot_type)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        next_state = state.copy()
+        x, y, is_pick, is_place = action.arr
```

Comment:
so this is going to do something interesting only when is_pick is True, is_place is True, or both are false, which corresponds to a fasten. right? do you want to `assert not (is_pick > 0.5 and is_place > 0.5)` for clarity?

Reply:
i'll make it a no-op instead of an assert, since an agent could plausibly pass in such an action (e.g., `random_actions_approach`). in general, asserts in simulate() should only occur when there's a literally impossible situation under the dynamics